### PR TITLE
[copy-operator] Add support for the copy operator in preparation for  making consuming and borrowing no implicit copy.

### PR DESCRIPTION
**Pull request by gottesmm**
rdar://101862423

### CODE REVIEW

**Comment by gottesmm** [comment]
https://github.com/apple/swift-syntax/pull/1669

@swift-ci test

File: test/SILGen/copy_expr.swift [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  var k = Klass()
+}
+
+// CHECK-LABEL: sil hidden [ossa] @$s9copy_expr22testCopyLoadableRValueyyF : $@convention(thin) () -> () {
+// CHECK: [[X:%.*]] = apply {{%.*}}({{%.*}}) : $@convention(method) (@thin ContainKlass.Type) -> @owned ContainKlass
+// CHECK: [[BORROW:%.*]] = begin_borrow [lexical] [[X]]
+// CHECK: [[COPY_BORROW:%.*]] = copy_value [[BORROW]]
+// CHECK: explicit_copy_value [[COPY_BORROW]]
+// CHECK: } // end sil function '$s9copy_expr22testCopyLoadableRValueyyF'
+func testCopyLoadableRValue() {
```
Comment (kavon) [inline test/SILGen/copy_expr.swift:16]: I think it's worth covering cases where the copy's result is passed to `consuming`, `borrowing`, and `inout` parameters to ensure the explicit copy is there. For example, `(copy x).prop` where `prop` has a `__consuming get`, or a regular `get`.

**Review by jckarter — APPROVED** [review]

File: lib/Parse/ParseExpr.cpp [hunk 2]
```diff
@@ -422,6 +422,20 @@ ParserResult<Expr> Parser::parseExprSequenceElement(Diag<> message,
     return sub;
   }
 
+  if (Tok.isContextualKeyword("copy") &&
+      peekToken().isAny(tok::identifier, tok::kw_self) &&
```
Comment (ahoppen) [inline lib/Parse/ParseExpr.cpp:426]: As I discussed offline with @gottesmm, I think this also needs to include dollar identifiers for anonymous closure arguments.

File: lib/Parse/ParseExpr.cpp [hunk 3]
```diff
@@ -422,6 +422,20 @@ ParserResult<Expr> Parser::parseExprSequenceElement(Diag<> message,
...
+      !peekToken().isAtStartOfLine()) {
+    Tok.setKind(tok::contextual_keyword);
+
+    SourceLoc copyLoc = consumeToken();
+    ParserResult<Expr> sub =
+        parseExprSequenceElement(diag::expected_expr_after_copy, isExprBasic);
+    if (!sub.hasCodeCompletion() && !sub.isNull()) {
+      sub = makeParserResult(new (Context) CopyExpr(copyLoc, sub.get()));
+    }
```
Comment (ahoppen) [inline lib/Parse/ParseExpr.cpp:435]: Why can’t we form a `CopyExpr` result if `sub` contains the code completion token?

Comment (gottesmm) [reply lib/Parse/ParseExpr.cpp:435]: I don't know. I think I was just copying what we already had. Should I allow for that?

File: test/SILGen/copy_expr.swift (hunk 1 above)
Comment (gottesmm) [reply test/SILGen/copy_expr.swift:32]: @kavon I added a bunch below.

File: lib/Parse/ParseExpr.cpp (hunk 2 above)
Comment (gottesmm) [reply lib/Parse/ParseExpr.cpp]: I fixed that.

File: lib/Parse/ParseExpr.cpp (hunk 3 above)
Comment (ahoppen) [reply lib/Parse/ParseExpr.cpp:436]: I don’t see a reason why we shouldn’t.

Comment (gottesmm) [reply lib/Parse/ParseExpr.cpp]: Thanks again for your help @ahoppen!

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
https://github.com/apple/swift-syntax/pull/1669

@swift-ci smoke test

**Comment by gottesmm** [comment]
@jckarter you probably want to take a quick additional look at this since I decided while I was here to also fix consume so it has the same behavior and also in the process fix another issue where consume was emitting proper errors for move only types, but due to an extra allocation that we introduced in the var case (due to us using rvalue logic instead of lvalue logic that the type checker work here allows us to), we would hit a copy of noncopyable type error. Now that we use the lvalue logic directly for consume like I do for copy, that is elided.

**Comment by gottesmm** [comment]
https://github.com/apple/swift-syntax/pull/1669

@swift-ci test

**Comment by gottesmm** [comment]
https://github.com/apple/swift-syntax/pull/1669

@swift-ci test

**Comment by gottesmm** [comment]
I updated the tests in lldb so that they work with these changes.

**Pull request merged**